### PR TITLE
core(lightwallet): compute resource metrics by origin

### PR DESCRIPTION
**Pull request by khempenius**
This PR is an intermediate step towards supporting a `firstPartyOrigins` configuration option in `budget.json`.

`computed/origin-summary.js` computes a breakdown, by origin, of resource metrics. This would be consumed by the `resource-summary` and `performance-budget` audits when providing statistics on third-party metrics.

### CODE REVIEW

File: lighthouse-core/computed/origin-summary.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+   */
+  static summarize(networkRecords) {
+    /** @type {Record<string,ResourceEntry>} */
+    const resourceSummary = {
+      'total': {count: 0, size: 0},
+    };
+
+    for (const record of networkRecords) {
+      const origin = URL.getOrigin(record.url);
+      if (origin) {
```
Comment (connorjclark) [inline lighthouse-core/computed/origin-summary.js]: `if (!origin) continue;`

File: lighthouse-core/computed/origin-summary.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/** @typedef {{count: number, size: number}} ResourceEntry */
+class OriginSummary {
+  /**
+   * @param {Array<LH.Artifacts.NetworkRequest>} networkRecords
+   * @return {Record<string,ResourceEntry>}
+   */
+  static summarize(networkRecords) {
+    /** @type {Record<string,ResourceEntry>} */
+    const resourceSummary = {
+      'total': {count: 0, size: 0},
```
Comment (connorjclark) [inline lighthouse-core/computed/origin-summary.js]:
```suggestion
      total: {count: 0, size: 0},
```

File: lighthouse-core/computed/origin-summary.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    };
+
+    for (const record of networkRecords) {
+      const origin = URL.getOrigin(record.url);
+      if (origin) {
+        if (resourceSummary[origin] === undefined) {
```
Comment (connorjclark) [inline lighthouse-core/computed/origin-summary.js]:
```suggestion
        if (!resourceSummary[origin]) {
```

File: lighthouse-core/test/computed/origin-summary.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    const result = await ComputedOriginSummary.request(artifacts, context);
+    assert.equal(result['http://example.com'].count, 1);
+    assert.equal(result['http://example.com'].size, 30);
+
+    assert.equal(result['https://example.com'].count, 1);
+    assert.equal(result['https://example.com'].size, 75);
+
+    assert.equal(result['http://subdomain.example.com'].count, 1);
+    assert.equal(result['http://subdomain.example.com'].size, 50);
+  });
```
Comment (connorjclark) [inline lighthouse-core/test/computed/origin-summary.js]: a test case with a `data:` url would be good too

File: lighthouse-core/computed/origin-summary.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
```
Comment (connorjclark) [inline lighthouse-core/computed/origin-summary.js]:
```suggestion
 * @license Copyright 2020 Google Inc. All Rights Reserved.
```

File: lighthouse-core/test/computed/origin-summary.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
```
Comment (connorjclark) [inline lighthouse-core/test/computed/origin-summary.js]:
```suggestion
 * @license Copyright 2020 Google Inc. All Rights Reserved.
```

File: lighthouse-core/test/computed/origin-summary.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    artifacts = mockArtifacts([
+      {url: 'http://example.com/file.html', resourceType: 'Document', transferSize: 30},
+      {url: 'https://foo.bar.com/script.js', resourceType: 'Script', transferSize: 50},
+      {url: 'http://foo.bar.com/file.jpg', resourceType: 'Image', transferSize: 70},
+    ]);
+    const result = await ComputedOriginSummary.request(artifacts, context);
+    assert.equal(result['http://example.com'].count, 1);
+    assert.equal(result['http://example.com'].size, 30);
+
+    assert.equal(result['http://cdn.example.com'].count, 2);
```
Comment (patrickhulce) [inline lighthouse-core/test/computed/origin-summary.js]: I feel like i'm missing something with this, shouldn't it be `foo.bar.com` and the next line should match?

File: lighthouse-core/test/computed/origin-summary.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    assert.equal(result['http://example.com'].size, 120);
```
Comment (patrickhulce) [inline lighthouse-core/test/computed/origin-summary.js]: also wait how in the world is this passing when `assert.equal(result['http://example.com'].size, 30);` passes above...

Comment (patrickhulce) [reply lighthouse-core/test/computed/origin-summary.js]: oooooh you need to rename this file to `origin-summary-test.js` so it will get run in CI :)

**Comment by khempenius** [comment]
Bumping this up for review.

**Pull request merged**